hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledExecutorContainer.java

```diff
@@ -234,48 +234,51 @@ public void promoteSuspended() {
                                          SplitBrainMergePolicy mergePolicy) {
         SerializationService serializationService = nodeEngine.getSerializationService();
-        serializationService.getManagedContext().initialize(mergingValue);
+        serializationService.getManagedContext().initialize(mergingEntry);
         serializationService.getManagedContext().initialize(mergePolicy);
 
-        // try to find an existing item with the same value
-        ScheduledTaskDescriptor match = null;
-        for (ScheduledTaskDescriptor item : tasks.values()) {
-            if (mergingValue.getValue().equals(item)) {
-                match = item;
+        // try to find an existing task with the same definition
+        ScheduledTaskDescriptor mergingTask = mergingEntry.getValue();
+        ScheduledTaskDescriptor existingTask = null;
+        for (ScheduledTaskDescriptor task : tasks.values()) {
+            if (mergingTask.equals(task)) {
+                existingTask = task;
                 break;
             }
         }
-
-        ScheduledTaskDescriptor merged;
-        if (match == null) {
-            // Missing incoming entry
-            merged = mergePolicy.merge(mergingValue, null);
-            if (merged != null) {
-                enqueueSuspended(merged, false);
+        if (existingTask == null) {
+            ScheduledTaskDescriptor newTask = mergePolicy.merge(mergingEntry, null);
+            if (newTask != null) {
+                enqueueSuspended(newTask, false);
+                return newTask;
             }
         } else {
-            // Found a match -> real merge
-            MergingValueHolder<ScheduledTaskDescriptor> existingValue = createMergeHolder(serializationService, match);
-            merged = mergePolicy.merge(mergingValue, existingValue);
-            if (merged != null && !merged.equals(match)) {
-                // Cancel matched one, before replacing it
-                match.cancel(true);
-                enqueueSuspended(merged, true);
-            } else {
-                merged = null;
+            MergingEntryHolder<String, ScheduledTaskDescriptor> existingEntry
+                    = createMergeHolder(serializationService, existingTask);
+            ScheduledTaskDescriptor newTask = mergePolicy.merge(mergingEntry, existingEntry);
+            if (newTask != null) {
+                boolean hasSameDescriptor = newTask.equals(existingTask);
+                boolean hasSameResult = existingTask.getTaskResult().equals(newTask.getTaskResult());
+                if (!hasSameDescriptor || !hasSameResult) {
+                    // cancel the existing task, before replacing it
+                    existingTask.cancel(true);
+                    enqueueSuspended(newTask, true);
+                    return newTask;
+                }
             }
         }
-
-        return merged;
+        // cancel the merging task, since it has not been merged
+        mergingTask.cancel(true);
```

Comment:
I'm not 100% sure if we have to do this. But since we do it for the `existingTask` if it gets replaced, I think it's good to cancel an ignored `mergingTask` as well.

Reply:
This is the serialized descriptor that arrived for merging. It was suspended at the source, just before collection, and will be disposed after the collection. `cancel` here will (should) have no effect, since this acts as a DTO at this stage.

Reply:
Okay, then I'll add this as comment and remove the cancel() call.